stdlib/public/SDK/Foundation/NSDictionary.swift

```diff
@@ -62,64 +62,99 @@ extension Dictionary : _ObjectiveCBridgeable {
   }
 
   public static func _forceBridgeFromObjectiveC(
-    _ d: NSDictionary,
+    _ x: NSDictionary,
     result: inout Dictionary?
   ) {
     if let native = [Key : Value]._bridgeFromObjectiveCAdoptingNativeStorageOf(
-        d as AnyObject) {
+        x as AnyObject) {
       result = native
       return
     }
 
     if _isBridgedVerbatimToObjectiveC(Key.self) &&
        _isBridgedVerbatimToObjectiveC(Value.self) {
-      result = [Key : Value](_cocoaDictionary: d)
+      result = [Key : Value](_cocoaDictionary: x)
       return
     }
 
-    if Key.self == String.self {
+    // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
+    // may not be backed by an NSDictionary.
+    var builder = _DictionaryBuilder<Key, Value>(count: x.count)
+    var maybeDict:Dictionary? = nil
```

Comment:
Hm, I think we should rather extend the builder abstraction to support potentially non-unique insertions (via a new variant of the `add` method) -- so that we can interleave known-unique insertions with potential duplicates, and still get builder's performance benefits for all unique keys.